### PR TITLE
lance dataset that overrides Dataset.scanner and Dataset.head

### LINKED ISSUE #141: Add convenience function for limit/offset
Right now to take advantage of Lance performance, we need to explicitly call `lance.scanner(..., limit=20, offset=10)`. It would be nice to be able to do something like `lance.dataset(uri).limit(n=20, offset=10)`.

Note that pyarrow's Dataset already has a `head` method and under the hood it's creating a scanner. BUT, it does not support Lance's scanner and limit/offset options. In addition, pyarrow's Dataset and Scanner are cython cdef classes so cannot be monkey patched.

Options:

1. Create Lance subclasses for FileSystemDatasetFactory and FileSystemDataset (both in C++ and cython) which returns the right scanner and supports more ScanOptions. We would then monkey-patch pyarrow.dataset._filesystem_dataset to use this new factory and dataset.
2. Create cython-only LanceFileSystemDataset with a factory method to create an instance from a FileSystemDataset. Then in `lance.dataset`, call this factory method so `lance.dataset` returns a Dataset that supports `Dataset.head(n=20, offset=10)`, which calls `lance.scanner` under the hood to support lance-specific ScanOptions.

### PR DESCRIPTION
**Pull request by changhiskhan**
closes #141

### CODE REVIEW

File: python/lance/_lib.pyx [hunk 1]
```diff
@@ -12,31 +12,38 @@ from pathlib import Path
 from pyarrow import Table
 
 from pyarrow._dataset cimport (
-    CDataset,
-    CFileWriteOptions,
-    CScanner,
-    Dataset,
-    FileFormat,
-    FileWriteOptions,
+CDataset,
```
Comment (eddyxu) [inline python/lance/_lib.pyx:15]: this indent is weird.

File: python/lance/_lib.pyx [hunk 2]
```diff
@@ -155,3 +162,147 @@ def WriteTable(table: Table,
...
+        """The FileFormat of this source."""
+        cdef FileFormat format = LanceFileFormat.__new__(LanceFileFormat)
+        format.init(self.filesystem_dataset.format())
+        return format
+
+    def head(self, n: int, offset: int = 0) -> Table:
+        scanner = self.scanner(limit=n, offset=offset)
+        return scanner.to_table()
+
+    def scanner(self, *args, **kwargs):
```
Comment (eddyxu) [inline python/lance/_lib.pyx:303]: Are we going to deprecate `lance.scanner()`? Could be just remove it from this PR.

File: python/lance/tests/test_api.py [hunk 3]
```diff
@@ -36,6 +36,16 @@ def test_simple_round_trips(tmp_path: Path):
     assert table == actual
 
 
+def test_head(tmp_path: Path):
```
Comment (eddyxu) [inline python/lance/tests/test_api.py:39]: can we add a test to make sure duckdb works with this dataset

**Review by eddyxu — COMMENTED** [review]
Minor nits

File: python/lance/tests/test_api.py (hunk 3 above)
Comment (changhiskhan) [reply python/lance/tests/test_api.py:39]: there is an existing unit test that runs a duckdb query (how i discovered the problem originally)

File: python/lance/_lib.pyx (hunk 2 above)
Comment (changhiskhan) [reply python/lance/_lib.pyx:303]: makes sense. Can just do `lance.dataset(...).scanner(...)`

Comment (changhiskhan) [reply python/lance/_lib.pyx:303]: fixed

File: python/lance/_lib.pyx (hunk 1 above)
Comment (changhiskhan) [reply python/lance/_lib.pyx]: fixed. not sure why pycharm is doing that

**Review by eddyxu — APPROVED** [review]

**Pull request merged**